src/main/java/net/minestom/server/utils/block/BlockBreakCalculation.java

```diff
@@ -0,0 +1,192 @@
+    // The vanilla client checks for bamboo breaking speed with item instanceof SwordItem.
+    // We could either check all sword ID's, or the sword tag.
+    // Since tags are immutable, checking the tag seems easier to understand
+    private static final Tag SWORD_TAG = Objects.requireNonNull(MinecraftServer.getTagManager().getTag(BasicType.ITEMS, "minecraft:swords"));
+
+    /**
+     * Calculates the block break time in ticks
+     *
+     * @return the block break time in ticks, -1 if the block is unbreakable
+     */
+    public static int breakTicks(@NotNull Block block, @NotNull Player player) {
+        if (player.getGameMode() == GameMode.CREATIVE) {
+            // Creative can always break blocks instantly
+            return 0;
+        }
+        // Taken from minecraft wiki Breaking#Calculation
+        // More information to mimic calculations taken from minecraft's source
+        Registry.BlockEntry registry = block.registry();
+        double blockHardness = registry.hardness();
+        if (blockHardness == -1) {
+            // Bedrock, barrier, and unbreakable blocks
+            return UNBREAKABLE;
+        }
+        ItemStack item = player.getItemInMainHand();
+        // Bamboo is hard-coded in client
+        if (block.id() == Block.BAMBOO.id() || block.id() == Block.BAMBOO_SAPLING.id()) {
+            if (SWORD_TAG.contains(item.material().namespace())) {
+                return 0;
+            }
+        }
+        Tool tool = item.get(ItemComponent.TOOL);
+        boolean isBestTool = canBreakBlock(tool, block);
+        float speedMultiplier;
+
+        if (isBestTool) {
+            speedMultiplier = getMiningSpeed(tool, block);
+
+            // Also, the wiki's canBreak seems to be this >1
+            if (speedMultiplier > 1F) {
+                // Wiki tells us about calculations for efficiency, but enchantments don't apply these calculations in minestom.
+                // If someone wants faster tools, they have to use player attributes or the TOOL component
+                speedMultiplier += (float) player.getAttributeValue(Attribute.PLAYER_MINING_EFFICIENCY);
+            }
+        } else {
+            speedMultiplier = 1;
+        }
+
+        if (player.hasEffect(PotionEffect.HASTE) || player.hasEffect(PotionEffect.CONDUIT_POWER)) {
+            // Yes, conduit power is same as haste. I also had to go confirm, because I couldn't believe it
+            speedMultiplier *= getHasteMultiplier(player);
+        }
+
+        if (player.hasEffect(PotionEffect.MINING_FATIGUE)) {
+            speedMultiplier *= getMiningFatigueMultiplier(player);
+        }
+
+        speedMultiplier *= (float) player.getAttributeValue(Attribute.PLAYER_BLOCK_BREAK_SPEED);
+
+        ItemStack helmet = player.getInventory().getHelmet();
+
+        if (isInWater(player) && !helmet.get(ItemComponent.ENCHANTMENTS, EnchantmentList.EMPTY).has(Enchantment.AQUA_AFFINITY)) {
```

Comment:
Do not check aqua affinity. The definition for the aqua affinity enchantment applies the submerged mining speed attribute.

```
    "minecraft:aqua_affinity": {
        effects: {
            "minecraft:attributes": [
                {
                    amount: {
                        base: 4b,
                        per_level_above_first: 4b,
                        type: "minecraft:linear"
                    },
                    attribute: "minecraft:player.submerged_mining_speed",
                    id: "minecraft:enchantment.aqua_affinity",
                    operation: "add_multiplied_total"
                }
            ]
        },
```